sinaloa/src/ec2_instance.rs

```diff
@@ -0,0 +1,317 @@
+            println!("ec2 result Stdout:{:}", ec2_result_text);
+            return Err(EC2Error::CLIError);
+        }
+
+        let ec2_result_text = std::str::from_utf8(&ec2_result.stdout)
+            .map_err(|err| EC2Error::Utf8Error(err))?;
+        let ec2_data: Value = serde_json::from_str(ec2_result_text)
+            .map_err(|err| EC2Error::SerdeJsonError(err))?;
+
+        let instance_id: &str = match &ec2_data["Instances"][0]["InstanceId"] {
+            Value::String(value) => value,
+            _ => return Err(EC2Error::IncorrectJson),
+        };
+        let private_ip: &str = match &ec2_data["Instances"][0]["PrivateIpAddress"] {
+            Value::String(value) => value,
+            _ => return Err(EC2Error::IncorrectJson),
+        };
+        println!("EC2Instance instance_id:{:?}", instance_id);
+        println!("EC2Instance private_ip:{:?}", private_ip);
+
+        std::thread::sleep(std::time::Duration::from_millis(30000));
+
+        let socket_port: u16 = 9090;
+
+        Ok(EC2Instance {
+            instance_id: instance_id.to_string(),
+            private_ip: private_ip.to_string(),
+            socket_port: socket_port,
+            socket_fd: None,
+        })
+    }
+
+    fn socket_connect(&mut self) -> Result<RawFd, EC2Error> {
+        let ip_addr: Vec<u8> = self.private_ip.split(".")
+            .map(|s| s.parse().expect("Parse error")).collect();
+        let inet_addr: InetAddr = InetAddr::new(IpAddr::new_v4(ip_addr[0], ip_addr[1], ip_addr[2], ip_addr[3]), self.socket_port);
+        let sockaddr = SockAddr::new_inet(inet_addr); 
+
+        // let socket_fd = socket(AddressFamily::Inet, SockType::Stream, SockFlag::empty(), None)
+        //     .expect("Failed to create socket");
+        let socket_fd = {
+            loop {
+                match socket(AddressFamily::Inet, SockType::Stream, SockFlag::empty(), None) {
+                    Ok(fd) => break fd,
+                    Err(nix::Error::Sys(err)) => {
+                        match err {
+                            nix::errno::Errno::ECONNREFUSED => {
+                                println!("EC2Instance::socket failed, ECONNREFUSED, trying again");
+                                continue
+                            },
+                            _ => panic!(format!("Failed to create socket:{:?}", err)),
+                        }
+                    },
+                    Err(err) => panic!(format!("Failed to create socket:{:?}", err)),
+                }
+            }
+        };
+        while let Err(_err) = connect(socket_fd, &sockaddr) {
+        }
+        //connect(socket_fd, &sockaddr).expect("Failed to connect to socket");
+
+
+        self.socket_fd = Some(socket_fd);
+
+        return Ok(socket_fd);
+    }
+
+    pub fn close(&mut self)-> Result<(), EC2Error> {
+        
+        if let Some(socket_fd) = self.socket_fd.take() {
+            match shutdown(socket_fd, Shutdown::Both) {
```

Comment:
Maybe clearer as `if let Err(err) = shutdown(socket_fd, Shutdown::Both) { eprintln!("..."); }`?

Reply:
Do you think so? I'm not so sure (other voices appreciated here @geky @ShaleXIONG @nspin @basmaelgaabouri )

Reply:
I don't think it really matters, `match` is common for error handling and `if let` is also common